### DISCUSSION TITLE
WHD02 switch help

### DISCUSSION THREAD

**Discussion by Ivano62** [Koenkk/zigbee2mqtt#12197] [Devices]
Hi, I've some whd02 switches, which seem to work fine.
I can't understand the difference between the three possible values that can be assigned to Switch_type (toggle, state, momentary), can anyone help me?
Thank you

**Comment by Ivano62** [comment]
Can anyone help me?

**Comment by Ivano62** [comment]
I have to control a lamp with a button, not a switch, and I can't figure out how to set the switch_type parameter (toggle, state or momentary). I am very grateful to anyone who can help me

**Reply by Serlli76** [reply]
I have the same question. I have tried each of the states separately and they worked the same way.

**Reply by diegocom** [reply]
Some question, any information about this?

**Reply by paulbusse** [reply]
It is explained [here](https://qscprod.force.com/selfhelpportal/s/article/What-are-the-differences-between-Toggle-Momentary-and-Trigger-buttons). In short:
- toggle: the switch is on until you turn it off (a light switch)
- momentary: the switch is on while you push on it (push to talk switch)
(examples stolen from the referenced article)

**Reply by chrisbraucker** [reply]
To add to that, imagine the smart switch as a device that has decoupled input (light switch) and output (light bulb) states.

- `toggle`: when you press the physical switch, the input state changes, and stays that way, reflecting the desired light state change (input state change == output state change). If you then interact with the smart switch via z2m, it remembers that the input and output switch state now should be reversed (important for `state`)
- `momentary`: the output is switched by a momentary change in the input, like a push-button that only closes the circuit when pushed and stays open otherwise. This means that every time the button is pressed-and-released, the state of the output should change. z2m interaction only changes the output state
- `state`: So `toggle` remembers to reverse the connection between input state and output state. This one doesn't. This mode connects the physical switch state to the light state and keeps it that way. (input = 1 <=> output = 1) So when you toggle the smart switch state digitally, you diverge from the physical state. In turn, the next time you toggle the physical switch, it may not change the state of the lights. I've not yet found a use case for this mode

**Reply by antonio1475** [reply]
> To add to that, imagine the smart switch as a device that has decoupled input (light switch) and output (light bulb) states.

Thank you for this detail, @chrisbraucker 

I have one of these switches connected to my doorbell, with the switch/pusher connected to S1 & S2, so that apart from working "offline" it lets Z2M know, so that I get a notification that the doorbell has been pressed.
This doorbell is a _ding-dong_ type, and when ON an electromagnet moves the bar to the left side to knock (_ding_), and when OFF the electromagnet turns off and the spring makes the bar knock on the right side (_dong_).

I've had it on Toggle for months without issue, but with your explanation I've switched to State. Now, if for whatever reason the switch is turned on digitally, the wall button will eventually restore it to its natural state (push won't do anything, release will turn it off).

In practice, the difference will be none, but I kind of like the "redundancy".
![image](https://user-images.githubusercontent.com/63130724/218308118-4efc454f-84fc-4f83-8d50-2666e1e6f1f1.png)


Momentary of course worked as you explained, but it's of no use for a doorbell (it will only trigger one state with each press&release, staying on or off indefinitely).

Just posting in case someone is in a similar doubt, providing the example of a doorbell.

👍

**Reply by countrr** [reply]
The State mode can be useful when reading the position of an output relay of for example an NVR, alarmsystem or thermostat. Any digital fiddling with the state will then not reverse the connection that you mentioned.

**Reply by Kopetan4egX** [reply]
> * I've not yet found a use case for this mode

Works perfect with a usual rocker switch. For example, without a relay, when the top part is pressed, the light is on, when bottom part - off. Using relay in "toggle" mode creates a situation, where light can be both on and off in both positions of switch,  depending on if you switched the relay digitally even or odd number of times. But with "state" mode switch is always pressed on top side with light on, and bottom - when off. The only drawback is a need to press switch additional time when digital and physical states don't match - light is off digitally, but top side is pressed, so you need to switch rocker down and back up to turn the light on and on reverse if light is on digitally.

**Reply by fnees** [reply]
> * `state`: So `toggle` remembers to reverse the connection between input state and output state. This one doesn't. This mode connects the physical switch state to the light state and keeps it that way. (input = 1 <=> output = 1) So when you toggle the smart switch state digitally, you diverge from the physical state. In turn, the next time you toggle the physical switch, it may not change the state of the lights. **I've not yet found a use case for this mode**

Just to add to the conversation, a use case for this mode are roller shades. In case it reversed the state it would turn the shades on when user is just turning the physical three state switch off. Thank you for the excelent explanation @chrisbraucker

**Comment by chrisbraucker** [comment]
@Ivano62 with regards to my reply just now, in order to switch a lamp with a momentary button that is connected to this smart switch, you should be able to set the state to `momentary` by selecting the device in zigbee2mqtt (opens the "About" tab), then navigate to the "Exposes" tab for that device, and there set the `switch_type` to `momentary`.

Navigate to Exposes:
<img width="985" alt="image" src="https://user-images.githubusercontent.com/7303810/215756054-776d27bd-a729-4c8b-95f6-1c925ef9b48b.png">

Then select `momentary`:
<img width="736" alt="image" src="https://user-images.githubusercontent.com/7303810/215756520-2d132c23-3dc0-4ca8-9b35-0539aa8f25ab.png">

**Comment by Ivano62** [comment]
Ok. Thanks
